app/app_test.go

```diff
@@ -261,275 +269,102 @@ func TestDifferentTxForms(t *testing.T) {
-		Fee:          0,
+	cases := []struct {
+		owner_index1   uint64
+		position1      types.Position
+		input_index1_1 uint64
+		input_index1_2 uint64
+		owner_index2   uint64
+		addr2          common.Address
+		position2      types.Position
+		input_index2_1 uint64
+		input_index2_2 uint64
+		newowner1      common.Address
+		denom1         uint64
+		newowner2      common.Address
+		denom2         uint64
+	}{
+		// 1 input 2 output
+		{0, types.NewPosition(0, 0, 0, 1), 0, 0, 0, common.Address{}, types.Position{}, 0, 0, addrs[1], 20, addrs[2], 80},
+		// 2 different inputs, 1 output
+		{1, types.NewPosition(7, 0, 0, 0), 0, 0, 2, addrs[2], types.NewPosition(7, 0, 1, 0), 0, 0, addrs[3], 100, common.Address{}, 0},
+		// 1 input 2 ouput
+		{3, types.NewPosition(8, 0, 0, 0), 1, 2, 0, common.Address{}, types.Position{}, 0, 0, addrs[3], 75, addrs[4], 25},
+		// 2 different inputs 2 outputs
+		{3, types.NewPosition(9, 0, 0, 0), 3, 0, 4, addrs[4], types.NewPosition(9, 0, 1, 0), 3, 0, addrs[3], 70, addrs[3], 30},
+		// 2 same inputs, 1 output (merge)
+		{3, types.NewPosition(10, 0, 0, 0), 3, 4, 3, addrs[3], types.NewPosition(10, 0, 1, 0), 3, 4, addrs[3], 100, common.Address{}, 0},
 	}
 
-	// Sign the hash of the transaction
-	hash := ethcrypto.Keccak256(msg.GetSignBytes())
-	sig1, _ := ethcrypto.Sign(hash, keys[1])
-	sig2, _ := ethcrypto.Sign(hash, keys[2])
-	tx = types.NewBaseTx(msg, []types.Signature{{
-		Sig: sig1,
-	}, {
-		Sig: sig2,
-	}})
-
-	cc.BeginBlock(abci.RequestBeginBlock{Header: abci.Header{Height: 8}})
-
-	// Run a check
-	cres = cc.Check(tx)
-	require.Equal(t, sdk.CodeType(0), sdk.CodeType(cres.Code), cres.Log)
+	for index, tc := range cases {
+		cc.BeginBlock(abci.RequestBeginBlock{Header: abci.Header{Height: 7 + int64(index)}})
+
+		msg := types.SpendMsg{
+			Blknum1:      tc.position1.Blknum,
+			Txindex1:     tc.position1.TxIndex,
+			Oindex1:      tc.position1.Oindex,
+			DepositNum1:  tc.position1.DepositNum,
+			Owner1:       addrs[tc.owner_index1],
+			ConfirmSigs1: CreateConfirmSig(tc.position1, keys[tc.input_index1_1], keys[tc.input_index1_2], true),
+			Blknum2:      tc.position2.Blknum,
+			Txindex2:     tc.position2.TxIndex,
+			Oindex2:      tc.position2.Oindex,
+			DepositNum2:  tc.position2.DepositNum,
+			Owner2:       tc.addr2,
+			ConfirmSigs2: CreateConfirmSig(tc.position2, keys[tc.input_index2_1], keys[tc.input_index2_2], true),
+			Newowner1:    tc.newowner1,
+			Denom1:       tc.denom1,
+			Newowner2:    tc.newowner2,
+			Denom2:       tc.denom2,
+			Fee:          0,
+		}
+
+		tx := GetTx(msg, keys[tc.owner_index1], keys[tc.owner_index2], !utils.ZeroAddress(msg.Owner2))
+
+		// Run a check
+		cres := cc.Check(tx)
+		require.Equal(t, sdk.CodeType(0), sdk.CodeType(cres.Code), cres.Log)
 
-	dres = cc.Deliver(tx)
-	require.Equal(t, sdk.CodeType(0), sdk.CodeType(dres.Code), dres.Log)
+		dres := cc.Deliver(tx)
 
-	ctx = cc.NewContext(false, abci.Header{})
+		require.Equal(t, sdk.CodeType(0), sdk.CodeType(dres.Code), dres.Log)
```

Comment:
All error msgs should include what testcase it failed on by using `index` in error string